File: mobile/src/lib/notifications.js
// ─────────────────────────────────────────────────────────────────────────────
// Reminder delivery: scheduling now lives server-side (Cloud Tasks for
// one-time reminders, a Cloud Scheduler sweep for recurring ones — see
// /functions) because on-device AlarmManager scheduling proved unreliable.
// This file only talks to that backend to schedule/cancel a push. Reminder
// text itself never leaves the device — only the reminder's own id (as an
// opaque notificationId) and its remind-at time/recurrence rule are sent;
// the push that eventually arrives carries a fixed generic alert, not the
// text (see /functions and push.js).
// ─────────────────────────────────────────────────────────────────────────────
import { LocalNotifications } from '@capacitor/local-notifications';
import { getOrCreateDeviceId } from './db.js';

const FUNCTIONS_BASE = 'https://asia-south1-track-blood.cloudfunctions.net';

export async function ensureNotificationPermission() {
  const { display } = await LocalNotifications.checkPermissions();
  if (display === 'granted') return true;
  const result = await LocalNotifications.requestPermissions();
  return result.display === 'granted';
}

// Read-only check (never prompts) — used to gate the reminder-creation UI.
// 'denied' means the user was already asked and said no; 'prompt' means they
// haven't been asked yet, which shouldn't block anything.
export async function getNotificationPermissionState() {
  const { display } = await LocalNotifications.checkPermissions();
  return display;
}

async function callFunction(name, payload) {
  const res = await fetch(`${FUNCTIONS_BASE}/${name}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`${name} failed: ${res.status}`);
  return res.json();
}

// Fires a local notification right now — used when a push arrives while the
// app is in the foreground (push.js), since the OS doesn't auto-display FCM
// alerts there. `notificationId`/`type` are stashed in `extra` so a tap on
// *this* notification (a separate tap-event system from FCM's, see push.js)
// can still be traced back to the reminder and handled the same way.
export async function showReminderNow(text, notificationId, type) {
  await LocalNotifications.schedule({
    notifications: [{
      id: Math.floor(Math.random() * 2_147_483_647),
      title: 'Track Blood',
      body: text,
      extra: { notificationId, type },
    }],
  });
}

export async function scheduleReminder(id, text, remindAtISO, recurrence) {
  const granted = await ensureNotificationPermission();
  if (!granted) return id;
  const deviceId = await getOrCreateDeviceId();
  await callFunction('scheduleReminder', {
    deviceId,
    notificationId: id,
    remindAt: remindAtISO,
    recurrence: recurrence ?? null,
  });
  return id;
}

export async function cancelReminder(id) {
  const deviceId = await getOrCreateDeviceId();
  await callFunction('cancelReminder', { deviceId, notificationId: id });
}

File: mobile_flutter/lib/services/notifications.dart
// Port of mobile/src/lib/notifications.js. Reminder scheduling lives
// server-side (Cloud Tasks/Scheduler, see /functions) — this only talks to
// that backend to schedule/cancel a push, and shows the local re-display
// notification when a push arrives while the app is in the foreground
// (see push.dart). Reminder text never leaves the device.
import 'dart:convert';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:http/http.dart' as http;
import 'package:permission_handler/permission_handler.dart';
import 'db.dart';

const _functionsBase = 'https://asia-south1-track-blood.cloudfunctions.net';

final FlutterLocalNotificationsPlugin localNotifications = FlutterLocalNotificationsPlugin();

Future<bool> ensureNotificationPermission() async {
  final status = await Permission.notification.status;
  if (status.isGranted) return true;
  return (await Permission.notification.request()).isGranted;
}

// Read-only check (never prompts) — used to gate the reminder-creation UI.
Future<PermissionStatus> getNotificationPermissionState() => Permission.notification.status;

Future<Map<String, dynamic>> _callFunction(String name, Map<String, dynamic> payload) async {
  final res = await http.post(Uri.parse('$_functionsBase/$name'), headers: {'Content-Type': 'application/json'}, body: jsonEncode(payload));
  if (res.statusCode != 200) throw Exception('$name failed: ${res.statusCode}');
  return jsonDecode(res.body) as Map<String, dynamic>;
}

// Fires a local notification right now — used when a push arrives while the
// app is in the foreground, since the OS doesn't auto-display FCM alerts there.
Future<void> showReminderNow(String text, String? notificationId, String? type) async {
  await localNotifications.show(
    id: DateTime.now().millisecondsSinceEpoch.remainder(2147483647),
    title: 'Track Blood',
    body: text,
    payload: jsonEncode({'notificationId': notificationId, 'type': type}),
    notificationDetails: const NotificationDetails(
      android: AndroidNotificationDetails('reminders', 'Reminders', importance: Importance.high, priority: Priority.high),
    ),
  );
}

Future<int> scheduleReminder(int id, String text, String remindAtIso, String? recurrence) async {
  final granted = await ensureNotificationPermission();
  if (!granted) return id;
  final deviceId = await Db.instance.getOrCreateDeviceId();
  await _callFunction('scheduleReminder', {'deviceId': deviceId, 'notificationId': id, 'remindAt': remindAtIso, 'recurrence': recurrence});
  return id;
}

Future<void> cancelReminder(int id) async {
  final deviceId = await Db.instance.getOrCreateDeviceId();
  await _callFunction('cancelReminder', {'deviceId': deviceId, 'notificationId': id});
}
